### PR TITLE
[Go] Add a generator for the gin framework

### PR DESCRIPTION
**Pull request by kemokemo**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.3.x`, `4.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@antihax @bvwells @grokify 

### Description of the PR

Issue #1047 

I respect the openapi-generator.
I would be pleased if you could give us your guidance.

### CODE REVIEW

**Comment by antihax** [comment]
I would try to condense the java changes into the main go server/abstract generator and only create a separate template for the Gin server; if this is at all possible.

**Comment by wing328** [comment]
IMO it's ok to extends Go Gin generator with Abstract Go class for day-1 requirement. If we see more Go server generators contributed by the community, we can refactor common code into a Abstract Go server class similar to what we've done for Java JAX-RS generators

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoGinServerCodegen.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018 OpenAPI-Generator Contributors (https://openapi-generator.tech)
+ * Copyright 2018 SmartBear Software
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoGinServerCodegen.java]: @kemokemo I believe we no longer need line 3 for new files added directly to this project (OpenAPI Generator). Please refer to https://opensource.stackexchange.com/a/7301/12997 for more information.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoGinServerCodegen.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+     * Returns human-friendly help for the generator.  Provide the consumer with help
+     * tips, parameters here
+     *
+     * @return A string value for the help message
+     */
+    @Override
+    public String getHelp() {
+        return "Generates a Go server library using OpenAPI-Generator. By default, " +
+                "it will also generate service classes -- which you can disable with the `-Dnoservice` environment variable.";
+    }
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoGinServerCodegen.java]: @kemokemo I think we will need to revise the help text, e.g. to mention about Go Gin framework.

Does this generator also support the `-Dnoservice` option?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoGinServerCodegen.java (hunk 1 above)
Comment (kemokemo) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoGinServerCodegen.java]: Thank you for your review and the information about the 'Copyright notice in the file header'.
I'll delete these 3 lines.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoGinServerCodegen.java (hunk 2 above)
Comment (kemokemo) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/GoGinServerCodegen.java]: Oh.. Sorry.
I'll make the help message more appropriate and delete the mention about `-Dnoservice`.

**Comment by kemokemo** [comment]
I had neglected confirmation. I'm sorry.
Thank you so much for your support!
I am proud that I was committed to OpenAPI Generator 😃

**Comment by wing328** [comment]
I did some tests and was able to run it directly and via Docker. Partial result below:
```
[GIN-debug] POST   /v2/user                  --> _/go/src/go.CreateUser (3 handlers)
[GIN-debug] POST   /v2/user/createWithArray  --> _/go/src/go.CreateUsersWithArrayInput (3 handlers)
[GIN-debug] POST   /v2/user/createWithList   --> _/go/src/go.CreateUsersWithListInput (3 handlers)
[GIN-debug] DELETE /v2/user/{username}       --> _/go/src/go.DeleteUser (3 handlers)
[GIN-debug] GET    /v2/user/{username}       --> _/go/src/go.GetUserByName (3 handlers)
[GIN-debug] GET    /v2/user/login            --> _/go/src/go.LoginUser (3 handlers)
[GIN-debug] GET    /v2/user/logout           --> _/go/src/go.LogoutUser (3 handlers)
[GIN-debug] PUT    /v2/user/{username}       --> _/go/src/go.UpdateUser (3 handlers)
[GIN-debug] Listening and serving HTTP on :8080
```

**Comment by wing328** [comment]
PR merged into master. Thank you for your contribution 👍

**Comment by wing328** [comment]
Reddit: https://www.reddit.com/r/golang/comments/9hs0h6/go_gin_server_generator_is_the_96th_generator_to/

Twitter: https://twitter.com/oas_generator/status/1043183765982302208

Hacker News: https://news.ycombinator.com/item?id=18041132

**Comment by kemokemo** [comment]
I am glad that I could contribute to a wonderful project!
Thank you very much! 👍

**Comment by wing328** [comment]
@kemokemo thanks again for the new generator, which is included in the v3.3.0 minor release: https://twitter.com/oas_generator/status/1046941449609068544

**Comment by kemokemo** [comment]
@wing328  Thank you so much for your support! I'm very happy 😄

**Pull request merged**